Git subdir

Fuuzetsu opened the pull request.

jkarni commented:
LGTM - the only minor quibble is that maybe a new datatype would be clearer/safer than `(String, Maybe String)`

sol commented on hpack.cabal:
```diff
@@ -31,6 +31,7 @@ executable hpack
     , deepseq
     , directory
     , filepath
+    , split
```
I think `text` also provides the functionality we need from `split` and is already a transitive dependency, so I would prefer to use that.

sol commented on test/RunSpec.hs:
```diff
@@ -102,6 +102,19 @@ spec = do
+        , "  type: git"
+        , "  location: https://github.com/hspec/hspec"
+        , "  subdir: driver"
+        ]
```
Do you think we could test `renderSourceRepository` independently?  I think the current testing approach does not scale.

sol commented on test/ConfigSpec.hs:
```diff
@@ -177,7 +177,16 @@ spec = around_ (inTempDirectoryNamed "foo") $ do
+        |]
+      Right (_, c) <- readPackageConfig "package.yaml"
+      packageSourceRepository c `shouldBe`
+        Just ("https://github.com/hspec/hspec", Just "driver")
```
If you change `driver` to `hspec-core` the example is more realistic ;).

sol commented on src/Config.hs:
```diff
@@ -159,7 +162,7 @@ data Package = Package {
 , packageLicenseFile :: Maybe FilePath
 , packageExtraSourceFiles :: [FilePath]
-, packageSourceRepository :: Maybe String
+, packageSourceRepository :: Maybe (String, Maybe String)
```
I think at this point I'm also in favor of adding a proper `SourceRepository` type (kudos to @jkarni).

Fuuzetsu commented:
Addressed the comments in new commits. Once everything is good I'll squash it of course.

Regarding `splitOn`, I replaced it with `Data.Text.split` as advised. There were multiple ways to do it: 

1. parse out Text from the yaml file, split what we get out and from there just use `String` so everything else works
2. Do the above but without converting to String. This made it a bit ugly as we were now mixing String and Text here and there.
3. Parse out as `String` but convert to `Text` just for the split.

I opted for 1. Option 3 would work too and would keep the `PackageConfig` field type unchanged if that's desired, very simple change. Option 2 could be nicer if we went ahead and refactored everything to just use `Text`.

sol commented on src/Config.hs:
```diff
@@ -186,6 +188,16 @@ data Executable = Executable {
+data GithubConfig = GithubConfig {
+  githubConfigUrl :: String
+, githubConfigSubdir :: Maybe String
+} deriving (Eq, Show)
```
Can we rename this type into `SourceRepository`?

sol commented on src/Config.hs:
```diff
@@ -261,21 +273,28 @@ mkPackage (CaptureUnknownFields unknownFields PackageConfig{..}) = do
-    github = ("https://github.com/" ++) <$> packageConfigGithub
+    github :: Maybe GithubConfig
+    github = modifyGithubUrl ("https://github.com/" <>) <$>
+             maybe Nothing (Just . splitGithub) packageConfigGithub
```
^ please don't align

sol commented on src/Config.hs:
```diff
@@ -261,21 +273,28 @@ mkPackage (CaptureUnknownFields unknownFields PackageConfig{..}) = do
+    github = modifyGithubUrl ("https://github.com/" <>) <$>
+             maybe Nothing (Just . splitGithub) packageConfigGithub
+      where
+        splitGithub str = case T.unpack <$> T.split (== '/') str of
```
Using `fmap` on lists I consider code obfuscation.  See also https://ro-che.info/articles/2015-02-12-foldable-traversable-parametricity by @feuerbach.

sol commented on src/Config.hs:
```diff
@@ -261,21 +273,28 @@ mkPackage (CaptureUnknownFields unknownFields PackageConfig{..}) = do
       _ -> join packageConfigHomepage <|> fromGithub
       where
-        fromGithub = ((++ "#readme") <$> github)
+        fromGithub = (<> "#readme") . githubConfigUrl <$> github
```
As we stick with `String` I would prefer to not change this to `<>` (as `<>` is not available in older versions of base).

sol commented on test/RunSpec.hs:
```diff
@@ -150,3 +138,18 @@ spec = do
           , "  ghc-options: -Wall -Werror"
           , "  default-language: Haskell2010"
           ]
+
```
The following specs should be group under `describe "renderSourceRepository"`

sol commented on test/RunSpec.hs:
```diff
@@ -150,3 +138,18 @@ spec = do
+        `shouldBe` unlines [ "source-repository head"
+                           , "  type: git"
+                           , "  location: https://github.com/hspec/hspec"
+                           ]
```
^ please don't align.  Use e.g. the following instead:

```
`shouldBe` unlines [
    "source-repository head"
  , "  type: git"
  , "  location: https://github.com/hspec/hspec"
  ] 
```

sol replied:
My reason is that if you change this code to e.g. `(T.pack . unlines) [...` later on, you have to realign the whole list.

sol commented:
@Fuuzetsu Yes, I think your approach regarding `String` and `Text` is good.  Other then my comments, LGTM.

Fuuzetsu commented:
See latest commit.
